From e5de96a92d677c60054da546b95090f3e9adade2 Mon Sep 17 00:00:00 2001
From: Yuan Juntao <juntao.yuan@intel.com>
Date: Thu, 13 Nov 2014 15:26:46 +0800
Subject: [PATCH] regulator: add vmmc & vqmmc regulators for SD host with
 whiskey pmic

Add VSDCARD and VSDIO regulators for SD host controller.

Change-Id: Iebd5d43368ac2bf643dfc79a2c7bca0207d67b03
Signed-off-by: Yuan Juntao <juntao.yuan@intel.com>
---
 .../device_libs/platform_sdio_regulator.c          | 36 ++++++++++++++++++++++
 drivers/mfd/intel_soc_pmic_wc.c                    | 13 ++++++++
 2 files changed, 49 insertions(+)

diff --git a/arch/x86/platform/intel-mid/device_libs/platform_sdio_regulator.c b/arch/x86/platform/intel-mid/device_libs/platform_sdio_regulator.c
index d7dd554..aa3a883 100644
--- a/arch/x86/platform/intel-mid/device_libs/platform_sdio_regulator.c
+++ b/arch/x86/platform/intel-mid/device_libs/platform_sdio_regulator.c
@@ -24,6 +24,7 @@
 #include <linux/mfd/intel_soc_pmic.h>
 #include <linux/platform_device.h>
 #include <asm/cpu_device_id.h>
+#include <linux/regulator/intel_whiskey_cove_pmic.h>
 
 struct acpi_ids { char *hid; char *uid; };
 
@@ -171,10 +172,45 @@ static void intel_setup_ccove_sd_regulators(void)
 			sizeof(struct fixed_voltage_config), 0);
 }
 
+/*************************************************************
+*
+* WCOVE SD card related regulator
+*
+*************************************************************/
+static struct regulator_init_data wcove_vmmc_data;
+static struct regulator_init_data wcove_vqmmc_data;
+
+static struct wcove_regulator_info wcove_vmmc_info = {
+	.init_data = &wcove_vmmc_data,
+};
+
+static struct wcove_regulator_info wcove_vqmmc_info = {
+	.init_data = &wcove_vqmmc_data,
+};
+
+static void intel_setup_whiskey_cove_sd_regulators(void)
+{
+	memcpy((void *)&wcove_vmmc_data, (void *)&vmmc_data,
+			sizeof(struct regulator_init_data));
+	memcpy((void *)&wcove_vqmmc_data, (void *)&vqmmc_data,
+			sizeof(struct regulator_init_data));
+
+	/* set enable time for vqmmc regulator, stabilize power rail */
+	wcove_vqmmc_data.constraints.enable_time = 20000;
+
+	/* register SD card regulator for whiskey cove PMIC */
+	intel_soc_pmic_set_pdata("wcove_regulator", &wcove_vmmc_info,
+		sizeof(struct wcove_regulator_info), WCOVE_ID_V3P3SD + 1);
+	intel_soc_pmic_set_pdata("wcove_regulator", &wcove_vqmmc_info,
+		sizeof(struct wcove_regulator_info), WCOVE_ID_VSDIO + 1);
+}
+
 static int __init sdio_regulator_init(void)
 {
 	intel_setup_ccove_sd_regulators();
 
+	intel_setup_whiskey_cove_sd_regulators();
+
 	return 0;
 }
 fs_initcall_sync(sdio_regulator_init);
diff --git a/drivers/mfd/intel_soc_pmic_wc.c b/drivers/mfd/intel_soc_pmic_wc.c
index ab4c43f..e91dda4 100644
--- a/drivers/mfd/intel_soc_pmic_wc.c
+++ b/drivers/mfd/intel_soc_pmic_wc.c
@@ -36,6 +36,7 @@
 #include <linux/mfd/intel_soc_pmic_wcove.h>
 #include <asm/intel_em_config.h>
 #include "./intel_soc_pmic_core.h"
+#include <linux/regulator/intel_whiskey_cove_pmic.h>
 
 #define WHISKEY_COVE_IRQ_NUM	17
 
@@ -406,6 +407,18 @@ static struct mfd_cell whiskey_cove_dev[] = {
 		.resources = NULL,
 	},
 	{
+		.name = "wcove_regulator",
+		.id = WCOVE_ID_V3P3SD + 1,
+		.num_resources = 0,
+		.resources = NULL,
+	},
+	{
+		.name = "wcove_regulator",
+		.id = WCOVE_ID_VSDIO + 1,
+		.num_resources = 0,
+		.resources = NULL,
+	},
+	{
 		.name = "whiskey_cove_region",
 	},
 	{NULL, },
-- 
1.9.1

